velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,215 @@
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(groups[row]);
+      } else {
+        computeHash(groups[row], rawHashes[row]);
+      }
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto vector = arg->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      *value<int64_t>(groups[row]) += rawValues[row];
+    });
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(group);
+      } else {
+        computeHash(group, rawHashes[row]);
+      }
+    });
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushDown*/) override {
+    const auto& arg = args[0];
+    auto vector = arg->asUnchecked<FlatVector<int64_t>>();
+    VELOX_CHECK(vector);
+    auto rawValues = vector->rawValues();
+    int64_t result = 0;
+    rows.applyToSelected([&](vector_size_t i) { result += rawValues[i]; });
+    *value<int64_t>(group) += result;
+  }
+
+ private:
+  FOLLY_ALWAYS_INLINE void computeHash(char* group, const int64_t hash) {
+    clearNull(group);
+    auto result = *value<int64_t>(group);
+    result += hash * PRIME64_1;
+    *value<int64_t>(group) = result;
+  }
+
+  FOLLY_ALWAYS_INLINE void computeHashForNull(char* group) {
+    clearNull(group);
+    *value<int64_t>(group) += PRIME64_1;
+  }
+
+  FOLLY_ALWAYS_INLINE PrestoHasher* getPrestoHasher(TypePtr typePtr) {
+    if (prestoHasher_ == nullptr) {
+      prestoHasher_ = std::make_unique<PrestoHasher>(std::move(typePtr));
+    }
+    return prestoHasher_.get();
+  }
+
+  FOLLY_ALWAYS_INLINE BufferPtr& getHashBuffer(
+      vector_size_t size,
+      velox::memory::MemoryPool* pool) {
+    // hashes_->size() is in bytes.
+    if (hashes_ == nullptr || hashes_->size() < (size * sizeof(int64_t))) {
+      hashes_ = AlignedBuffer::allocate<int64_t>(size, pool);
```

Comment:
use AlignedBuffer::reallocate if hashes_ is not null; in case there is sufficient capacity in the buffer, reallocate will reuse memory instead of allocating new one.